### DISCUSSION TITLE
Performance slow when assign values in a loop

### DISCUSSION THREAD

**Discussion by zliucd** [nlohmann/json#3124] [Q&A] [answered]
Hi all,

The performance was very slow when assigning values in a loop(3000 times), and each time a new` ordered_json` string is generated and returned.

Can any give me a hint to accelerate?  thanks.

**Comment by nlohmann** [comment, marked as answer]
- Which compiler are you using?
- Are you running in debug or release mode?
- Do you have a concrete code example?

**Reply by zliucd** [reply]
Thanks for quick reply. 

- Compiler is Clang 12 in debug mode.
-  I'm using Valgrind callgrind for profiling, and statistics show `nlohmann::ordered_map` used (surprisingly ) many instruction cycles.
- When I switched from `ordered_json` to `json`, performance became slightly better (though not as fast I expected).

**Reply by nlohmann** [reply]
- Don't expect miracles in debug mode - the library checks a lot of assertions with every method call.
- Again: what are you executing in detail that surprise you in the instruction count?
- The difference between `ordered_json` and `json` is the way JSON objects are stored. `json` uses a `std::map`, whereas `ordered_json` uses a `std::vector` of `std::pair` and some bookkeeping to maintain the insertion order.

**Reply by zliucd** [reply]
I checked the code again; it turned in out many-times format string in a loop is the real cause.  I tried many int-to-string conversion implementations(e.g., ```{fmt}, snprintf, to_string()```), and they are all slow.

Example code:

```cpp
uint32_t len= 10000;
xx[len] = ...                                 // holds values

ordered_json j;
for(int i = 0; i < len; i++){
    string s = "abc_" + convert_int_to_string(i);   // 10x slower than not using string conversion.
    j[s] = xx[i];
}
```

**Reply by nlohmann** [reply]
- What type is `xx`?
- Is the time taken in the construction of the string from integer `i` or by the line below?

**Reply by zliucd** [reply]
`xx` is a `uint_16` array, and time is computed for the whole `for` loop scope.    The tricky thing here is `xx` represents a data feature after one hot encoding, so it has multiple `columns`(indexes).  The json string `j` will be passed to inline Python code, and it is ok pass raw data to Python for parsing column encoding, but if could be done C++, it's more clean.

Maybe we should avoid formatting string in a loop..

**Reply by nlohmann** [reply]
You mean code like this?

```cpp
#include <iostream>
#include <nlohmann/json.hpp>
#include <algorithm>

using ordered_json = nlohmann::ordered_json;

int main() {
    constexpr uint32_t len = 10000;
    std::array<std::uint16_t, len> xx{};
    for (auto i = 0; i < len; ++i) {
        xx[i] = i;
    }

    ordered_json j;
    for (int i = 0; i < len; i++) {
        std::string s = "abc_" + std::to_string(i);
        j[s] = xx[i];
    }

    std::cout << j["abc_999"] << std::endl;
}
```

On my machine, this runs in 2.59s (Debug) and 0.14s (Release).

If I replace `ordered_json j` with `std::map<std::string, uint32_t> j`, the Release run takes 0.01s, and I get the same runtime for `nlohmann::json j`.

**Reply by zliucd** [reply]
Yes, thanks for reproducing.  The performance between `ordered_json` and` json `is trivial, and the bottleneck is caused by lots of string conversion, and I end up not using string conversion in loop to avoid substantial perf loss.

**Reply by nlohmann** [reply]
So basically this issue can be closed as it's not related to the library as the performance was spent outside of it?

**Reply by zliucd** [reply]
I can't agree more.
